Implement MultiThreadedCache.jl

NHDaly opened the pull request:
Implements a MultiThreadedCache{K,V}, and adds stress test.

This cache has no locks on access, and only has contention on a cache
miss. It only ever holds a shared lock for a constant-time duration,
never while executing user code. A Task requesting a key that is
already being computed on another Task will block while that computation
is being performed. By taking advantage of the append-only properties of
a cache, the cache can be duplicated per-Thread, to avoid locking in the
common case.

This PR adds:
- implementation
- tests
- CI configuration

NHDaly committed:
Initial commit (via PkgTemplates)

NHDaly committed:
Add CI.yml - test with singlethreaded and multithreaded

NHDaly committed:
Finish implementing MultiThreadedCache, and add stress test.

This cache has no locks on access, and only has contention on a cache
miss. It only ever holds a shared lock for a constant-time duration,
never while executing user code. A Task requesting a key that is
already being computed on another Task will block while that computation
is being performed. By taking advantage of the append-only properties of
a cache, the cache can be duplicated per-Thread, to avoid locking in the
common case.

NHDaly committed:
Free the Futures once they're no longer needed, to prevent leaks

NHDaly commented:
PR Review Request: @tveldhui, @vilterp, @sacha0, @comnik ❤️

NHDaly committed:
Fix CI.yml syntax for environment variables

NHDaly committed:
Add julia "1" to build config

NHDaly committed:
Add MacOS and Windows to build config

kpamnany commented on test/MultiThreadedCaches.jl:
```diff
@@ -0,0 +1,76 @@
+    len = 100
+    outputs = [Channel{Int}(Inf) for _ in 1:len]
+
+    @sync for i in 1:100_000
+        Threads.@spawn begin
```
You have a correctness test here, but there's no perf test. I suggest using `@elapsed` and running this stress test twice; once with just 1 thread (no `@spawn`) and once with multiple threads. The comparison will show scalability.

NHDaly replied:
yeah, good idea. Should that go in `test`? Or should we have a separate benchmarks directory?

kpamnany replied:
Unless you plan on expanding the feature set dramatically, there's probably not much scope for a lot more tests or benchmarks? I'd put it here for now and split it out later if needed, but it's your call. :)

NHDaly replied:
Actually, Kiran, there's no scaling benefit here to running in multiple threads...

The scaling benefit comes from _caching_, not from parallelism here. The thread-safety requirement comes from the fact that you want to use those cached values even across multiple threads. The goal of this package is to provide that thread safety without adding too much contention that would _reduce_ performance as you scale. 

If everything is working well here, this test should have close to the same performance as running single-threaded, but certainly wouldn't be any faster. Does that make sense?

kpamnany replied:
Yes, what you want to verify is that the performance is good even in the presence of contention.

NHDaly replied:
gotcha. It seems to be, and I'll just have the benchmark print out the results. 👍 Thanks

NHDaly replied:
Here's a PR for the benchmark numbers: https://github.com/NHDaly/MultiThreadedCaches.jl/pull/4

NHDaly committed:
Add constructor that provides pre-computed values to the base_cache

tveldhui commented on src/MultiThreadedCaches.jl:
```diff
@@ -0,0 +1,103 @@
+            end
+        end
+        if is_first_task
+            v = func()
```
Handle case where `func()` throws an exception?

NHDaly replied:
oops, yes thanks

NHDaly replied:
🤔 I think that in Arroyo we cache the exception and throw it from now on, but since this open source package might be used in a more development-interactive environment, i'm going to just rethrow the exception and not cache anything, so that you can fix the exception and continue on.

Ideally no one is using exception throwing as a standard part of their code flow, since exceptions are expensive.

But thanks, i'll do the requisite clean up in this case 👍

NHDaly replied:
@tveldhui thanks, implemented here: https://github.com/NHDaly/MultiThreadedCaches.jl/pull/3

NHDaly committed:
Merge pull request #2 from NHDaly/nhd-constructors

Add constructor that provides pre-computed values to the base_cache

NHDaly committed:
Gracefully handle exceptions thrown during `get!()` functions

Rethrow the exception onto the Future for all blocked tasks, delete the
future, and rethrow on the current task.

The cache remains usable afterwards and nothing is recorded for the key
with the error.

NHDaly committed:
Add benchmark test measuring parallel scaling.

The results are a bit lukewarm (glad we measured it), but look alright.

Here's a few runs, each with different JULIA_NUM_THREADS:
```julia
┌ Info: benchmark results
│   Threads.nthreads() = 1
│   time_serial = 0.013336288
│   time_parallel = 0.09071632
└   time_baseline = 0.115363526
```
```julia
┌ Info: benchmark results
│   Threads.nthreads() = 2
│   time_serial = 0.011262138
│   time_parallel = 0.097021203
└   time_baseline = 0.139031655
```
```julia
┌ Info: benchmark results
│   Threads.nthreads() = 20
│   time_serial = 0.011997677
│   time_parallel = 0.658225544
└   time_baseline = 1.032283809
```
```julia
┌ Info: benchmark results
│   Threads.nthreads() = 100
│   time_serial = 0.013902211
│   time_parallel = 1.999772731
└   time_baseline = 9.314424419
```

So it definitely does not scale as well as a single-threaded codebase
would have scaled. But it also definitely scales better than the
baseline, which is a Mutex around a Dict.

Maybe there will be some places to improve our contention in the future :)

NHDaly committed:
Merge pull request #3 from NHDaly/nhd-exception-handling

Gracefully handle exceptions thrown during `get!()` functions

NHDaly commented:
Also i'm open to other package naming suggestions if anyone has them! :)

vilterp commented on src/MultiThreadedCaches.jl:
```diff
@@ -0,0 +1,177 @@
+
+    base_cache::Dict{K,V}  # Guarded by: base_cache_lock
+    base_cache_lock::ReentrantLock
+    base_cache_futures::Dict{K,Channel{V}}  # Guarded by: base_cache_lock
```
so channels are idiomatic for futures in Julia?

vilterp replied:
so channels are idiomatic for futures in Julia? I guess that makes sense

vilterp commented on src/MultiThreadedCaches.jl:
```diff
@@ -0,0 +1,177 @@
+                delete!(cache.base_cache_futures, key)
+            end
+            # Return v to any other Tasks that were blocking on this key.
+            put!(future, v)
```
will this send the value to all the tasks waiting on this channel, or just the first one? maybe it works because you're using `fetch`, not `take!`?

NHDaly replied:
yeah exactly, it's because i'm calling `fetch` not `take!`.

I'm not sure if it's exactly well-specified anywhere, but i've found that `Channel(1)` + `fetch()` makes for a Future. 👍

vilterp reviewed (approved):
LGTM with a couple questions

NHDaly committed:
Fix thread-safety violation by accidentally not locking the base cache during delete!

NHDaly committed:
Fix concurrency-safety by not holding the get!() on the thread-local cache across the whole execution

NHDaly committed:
Fix lazy construction of Dicts, per guidance from Julia Base

NHDaly committed:
Merge pull request #4 from NHDaly/nhd-benchmark-parallel-scaling

Add benchmark test measuring parallel scaling.

NHDaly committed:
Merge pull request #5 from NHDaly/nhd-fix-lazy-construction

Fix lazy construction of Dicts, per guidance from Julia Base

NHDaly committed:
Use per-thread locks to make invariant to Task migration.

...... In retrospect, this does make this whole structure start to look
a lot like a Dict + a Read/Write lock, and I wonder how their
performance would compare......

NHDaly committed:
Merge pull request #6 from NHDaly/nhd-concurrency-safety-fixes

Concurrency safety fixes

NHDaly commented:
After all the latest changes, especially #6 for the concurrency fixes, I think this package is good to go.

It still scales a good bit better than a Dict + Mutex, so I think there's still value in it!

If we want to give another approach a shot in the future, like a concurrent hash table, i'm super supportive. But hopefully this is useful in the interim.
